Configuring and starting Seastar

kfernandez31 opened the pull request:
This PR introduces the `AppTemplate` and `Options` structs as described in #1. To call Seastar's `app_template::run` under the hood, one must do so through `AppTemplate::run_{void, int}`, which accept function pointers. To make them accept an `async` block, one can do:

```rust
let opts = Options::default;
// set some options here
let app = AppTemplate::new_from_options(opts);
app.run_void(
    futures::executor::block_on(async { 
            some_async_work().await 
        }
);
```

## Notes:
- each `seastar-rs`-based app must have the above code, and so it'd be convenient to turn it into a macro. Luckily, we have an issue for that.
- each call of `AppTemplate::run_{void, int}` spawns a new thread. We might want to remove that slight overhead somehow.

Commit:
Added Options (+getting/setting), AppTemplate (+creating from Options)

Commit:
Initial code for app.run_{int, void}

Commit:
Added missing namespace

Commit:
Made num_cpus a test-only dep, changed return types of string getters to string&

Commit:
Added config_and_start_seastar.rs to bridges list, fixed some compilation errors

Commit:
Moved C++ type declarations to unsafe extern block, added missing references to function defs in config_and_start_seastar.cc, added const qualifiers

Commit:
Changed string getters to operate on &str

Commit:
Merge remote-tracking branch 'origin/api_level_config' into config_and_start_seastar

Commit:
Changed description getter to &str as well

Commit:
Added missing file() call to build.rs

Commit:
Added missing call of get_value

Commit:
Added proper handling for default smp

patjed41 committed:
Add AppTemplate::run_void and AppTemplate::run_int

sylwiaszunejko committed:
Made run_{void,int} spawn a new thread, added documentation and examples

Commit:
Addded public visibility for AppTemplate & Options, impl Default for AppTemplate, applied clippy and fmt

piodul commented:
> ```rust
> app.run_void(
>     futures::executor::block_on(async { 
>             some_async_work().await 
>         }
> );
> ```

This won't work.

`futures::executor::block_on` is a function that does not return until the rust future provided to it becomes ready. It basically steals the whole thread just for this particular rust future. This is bad because the runtime thread maintains a queue of tasks to execute and has to do some maintenance regularly, i.e. check timers and poll for I/O - so this prevents other tasks from being run and I/O from working correctly. In seastar, each task is supposed to execute for a small amount of time and then yield control to the runtime so that it gives a chance for other things to be done.

The original `app_template::run` takes a function which returns a `seastar::future<>`. You should use `cxx_async` in order to "convert" the rust future to a seastar future.

I guess the call to `app_template::run` should look a bit like this:

```c++
// `main_rust_future` is of future type generated by cxx_async
exit_value = app->run(ac, av, [&] () -> seastar::future<> {
    co_await main_rust_future;
});
```

On the other hand, the `run_void function should just take a future that does not return anything when resolved:

```rust
pub fn run_void(&self, args: &Vec<String>, fut: impl Future<Output=cxx_async::CxxAsyncResult<()>> + 'static) -> i32 {
    // `VoidFuture` type was generated by cxx_async
    // Perhaps you will have to work around the fact here that VoidFuture requires the future to be Send
    run_void(&self.app, args, VoidFuture::fallible(fut))
}
```

piodul commented on seastar/build.rs:
```diff
@@ -43,6 +44,7 @@ fn main() {
         .flag_if_supported("-fcoroutines")
         .includes(&seastar.include_paths)
         .cpp_link_stdlib("stdc++")
+        .file("src/config_and_start_seastar.cc")
```
How about a global static such as CXX_BRIDGES that will specify all the cpp files that the project needs to compile?

kfernandez31 replied:
Sure. In fact, such a variable already exists. It just wasn't used here.

piodul commented on seastar/src/config_and_start_seastar.cc:
```diff
@@ -0,0 +1,110 @@
+#include "config_and_start_seastar.hh"
+
+namespace seastar {
```
Please do not put things directly into the `seastar` namespace. You are risking name conflicts with items that might be introduced later by seastar. For example, `seastar_ffi` might be more appropriate.

Besides, lots of functions have names that are very likely to collide with items that we might try to introduce later (for example: `set_name`, `set_smp`). As `seastar_ffi` will be a root namespace of the project, I'm not sure if it is a good idea to put them there directly. I would introduce another namespace for those items (e.g. `seastar_ffi::configure_and_start_seastar`).

kfernandez31 replied:
Alright.

piodul commented on seastar/src/config_and_start_seastar.cc:
```diff
@@ -0,0 +1,110 @@
+
+// Copies rust::Vec<rust::String> as char**. Returns nullptr if fails.
+// Function free_args should be called on returned pointer to avoid memory leak.
+static char** args_as_ptr(const rust::Vec<rust::String>& args) {
+    char** av = (char**) calloc(args.size() + 1, sizeof(char*));
+    if (av == nullptr) {
+        return nullptr;
+    }
+
+    for (size_t i = 0; i < args.size(); i++) {
+        const rust::String& arg = args[i];
+        av[i] = (char*) calloc(arg.size() + 1, sizeof(char));
+        if (av[i] == nullptr) {
+            return nullptr;
+        }
+        strncpy(av[i], arg.data(), arg.size());
+    }
+
+    return av;
+}
```
Why are you using calloc? We are using C++. Why not use `new char*[args.size() + 1]` etc.?

IMO we should use Rust to convert the string arguments to the right representation and then pass it to C++. It should be shorter and less prone to memory leaks.

kfernandez31 replied:
Sure, I can use `operator new`. However, I'm not convinced about the second part, especially after the memory leak described in the other comment will be resolved. I tried to do it in Rust but it wasn't as pleasant.

piodul commented on seastar/src/config_and_start_seastar.hh:
```diff
@@ -0,0 +1,30 @@
+
+void set_description(const std::unique_ptr<seastar_options>& opts, const rust::Str description);
+
+void set_smp(const std::unique_ptr<seastar_options>& opts, const uint32_t smp);
```
Why not pass `seastar_options&`?

kfernandez31 replied:
Rust doesn't have access to that type directly, as it's opaque. It gets it under a `UniquePtr`, so it must transfer it through a `UniquePtr` as well.

kfernandez31 replied:
Indeed, that'd even make for shorter code.

piodul commented on seastar/src/config_and_start_seastar.rs:
````diff
@@ -0,0 +1,293 @@
+    ///
+    /// assert_eq!(app.run_void(&args, func), 0);
+    /// ```
+    pub fn run_void(&self, args: &Vec<String>, func: fn()) -> i32 {
````
Instead of passing a reference to a vector, it's slightly better to pass a slice. It allows for more use cases (you can easily convert a `&Vec` to `&[]`, but not the other way around.

Same suggestion applies to `run_int`.

kfernandez31 replied:
Agree.

piodul commented on seastar/src/config_and_start_seastar.cc:
```diff
@@ -0,0 +1,110 @@
+}
+
+void set_description(const std::unique_ptr<seastar_options>& opts, const rust::Str description) {
+    opts->description = std::string(description.begin(), description.size());;
```
Nit: double semicolon at the end of the line

kfernandez31 replied:
Must have been my Ocaml muscle memory. I just noticed, there should be `seastar::sstring`, not `std::string` too.

piodul commented on seastar/src/config_and_start_seastar.cc:
```diff
@@ -0,0 +1,110 @@
+        const rust::String& arg = args[i];
+        av[i] = (char*) calloc(arg.size() + 1, sizeof(char));
+        if (av[i] == nullptr) {
+            return nullptr;
```
Oops, you are leaking memory here. What about the strings that you have already allocated and the `av` array?

kfernandez31 replied:
Yep, thanks.

piodul commented on seastar/src/config_and_start_seastar.rs:
```diff
@@ -0,0 +1,293 @@
+        let opts = Options::new();
+        assert_eq!(opts.get_name(), "App");
+        assert_eq!(opts.get_description(), "");
+        assert_eq!(opts.get_smp(), num_cpus::get() as u32);
```
Are `opts.get_smp()` and `num_cpus::get()` guaranteed to return the same on all systems? Do they return exactly the same thing?

kfernandez31 replied:
Seastar:
```cpp
/// Configuration for the multicore aspect of seastar.
struct smp_options : public program_options::option_group {
    /// Number of threads (default: one per CPU).
    program_options::value<unsigned> smp;
 ...
 ```
 `num_cpus::get`: "This function will get the number of logical cores. [...]"
Due to Seastar's thread-per-core model, I should use `num_cpus::get_physical`, right?

Commit:
Fixed issues described in PR

piodul committed:
add seastar_test_guard module

Although seastar seems to support starting and stopping the runtime
multiple times in a sequence, running multiple runtimes in parallel in
the same process doesn't work. In order to support parallel unit tests,
a module is introduced that provides mutual exclusion for tests that use
the seastar runtime.

The `acquire_guard_for_seastar_test` function should be called by each
test which creates a seastar runtime, and the result held until the test
finishes. In the future when we get the seastar::test macro, this
function should be called transparently by the code generated by the
macro.

piodul committed:
gitignore: add .idea directory

piodul committed:
add cxx_async_local_future module

We are using cxx_async for passing futures/coroutines between Rust and
C++. The library was clearly made with work-stealing schedulers in mind,
so the Rust side imposes a `Send` requirement on the futures passed to
C++. However, seastar has a shared-nothing model so futures spawned on
tasks won't migrate between threads, therefore the `Send` requirement is
too much.

We are working around this limitation by lying to cxx_async - we tell it
that our futures are indeed `Send` even if they don't (and frequently
won't). It's technically speaking unsound because now the _safe_ code
needs to be careful not to accidentally pass a cxx_async wrapper future
to another thread - however, it's not super bad if we only use the
wrapper futures within the internals of seastar-rs. In the future, we
might improve the safety of the interface further, which would probably
require writing procedural macros that piggy-back on cxx_async's macros
but specify different bounds for the futures being wrapped.
Alternatively, we can investigate the possibility of dropping cxx_async
and writing a bespoke library that translates Rust futures to seastar
futures.

This commit introduces the cxx_async_local_future module which defines
an extension trait for futures generated by cxx_async and allows putting
non-sync futures into cxx_async futures.

piodul committed:
add config_and_start_seastar module

This commit introduces the configure_and_start_seastar module. As the
name suggests, it provides functionality necessary to configure and
start the seastar runtime.

It contains definitions of the following types:

- `Options` - wrapper over `seastar::app_template::options`. Provides
  configuration for the seastar runtime.
- `AppTemplate` - wrapper over `seastar::app_template`. Used to start
  the seastar runtime.

Example:

    let mut app = AppTemplate::default();
    let args = vec!["test"];
    let fut = async { Ok(42) };
    assert_eq!(app.run_int(&args[..], fut), 42);

Co-Authored-By: Kacper Fernandez <kacper.fernandez@gmail.com>
Co-Authored-By: sylwiaszunejko <sylwia.szunejko@gmail.com>
Co-Authored-By: Patryk Jędrzejczak <patryk.jed123@gmail.com>

sylwiaszunejko committed:
Merge branch 'zpp-2022-rust-seastar:main' into config_and_start_seastar

Commit:
Merge configure-runtime-redux into config_and_start_seastar

piodul commented:
This PR has been superseded by https://github.com/zpp-2022-rust-seastar/seastar-rs/pull/23, closing.
